Feature/vs2961/backend-auth

Anthonyp0329 opened the pull request:
### Tracking Info

Resolves #<issue number>

Make sure your branch name conforms to: `<feature|staging|bugfix|...>/<username>/<3-4 word description separated by dashes>`. Otherwise, please rename your branch and create a new PR.

### Changes

What changes did you make?

Created auth pages, with files to make calls to backend to register users. Sign in with google button needs to be implemented, and login page css fledged out. 

### Testing

How did you confirm your changes worked? 

Tested signup page css by resizing window, made sure users were registered in firebase console and mongo database. 

### Confirmation of Change 

Upload a screenshot, if possible. Otherwise, please provide instructions on how to see the change.

<img width="1440" alt="Screenshot 2023-02-19 at 1 07 34 PM" src="https://user-images.githubusercontent.com/76824702/219975311-661aa7fc-3d55-4d9c-8925-72a345d554ec.png">

vs2961 committed:
Add frontend auth files

Anthonyp0329 committed:
Login, signup setup

Put in login and signup page, linked to firebase auth

Anthonyp0329 committed:
Merge branch 'main' into Feature/vs2961/backend-auth

Anthonyp0329 committed:
SignUp Changes

Set up user register to store display name as well

vs2961 committed:
Write user creation functions

aammya8 committed:
Styling for Signup page

aammya8 committed:
Merge branch 'Feature/vs2961/backend-auth' of https://github.com/TritonSE/SDCTA-Data-Visualization-Website into Feature/vs2961/backend-auth

Anthonyp0329 committed:
Setting up backend calls

auth.tsx is file containing function calls to backend

Anthonyp0329 committed:
Merge branch 'Feature/vs2961/backend-auth' of https://github.com/TritonSE/SDCTA-Data-Visualization-Website into Feature/vs2961/backend-auth

Anthonyp0329 committed:
registerUser

Getting register post request to backend setup

vs2961 committed:
Add create user endpoint

Anthonyp0329 committed:
SignUpPage error handling + confirm password field

Added a confirm password field, as well as error handling for any sign up errors, such as the password not being long enough, the terms and services checkbox not being checked, email not being in use, or full name not being typed in. Added an error_handling folder with an auth-errors file used to handle the errors that arise when trying to register a user.

Anthonyp0329 committed:
LogIn page error handling

added error handling for login page, such as wrong password or user not found

aammya8 committed:
Setup Signup With Google

Setup Signup With Google using Firebase. Added google sign in button to signup form. CSS Styling still required.

aammya8 committed:
Cleaned up styling

Anthonyp0329 committed:
Squashed commit of the following:

commit 3f342fd21adfb6e24682779014fb1c8064e6c7a7
Author: Samvrit Srinath <114899512+SamvritSrinath@users.noreply.github.com>
Date:   Tue Feb 7 22:18:50 2023 -0800

    Feature/SamvritSrinath/Landing-Page (#10)

    * Basic React for Landing Page

    Used JSX(will return to EM for feedback)

    * modified:   ../.DS_Store
    	modified:   package-lock.json
    	modified:   package.json
    	modified:   src/App.jsx
    	modified:   src/components/Body.jsx
    	modified:   src/components/NavBar.jsx
    	src/Images/

    * new file:   SDCTA/src/Images/Graph.jpg
    	new file:   SDCTA/src/Images/SDTEF Logo Transparent Background 1.jpg
    	new file:   SDCTA/src/Images/unsplash_dsV4yvL5UJk.jpg

    * changed jsx to tsx

    * added separate css file for body component

    * modified:   src/App.css
    	modified:   src/App.tsx
    	new file:   src/components/Footer.css
    	modified:   src/components/Footer.tsx
    	deleted:    src/components/NavBar.jsx

    * Images in Body

    * Added Education Images
    	modified:   src/components/Body.tsx

    * changed heading

    * Updating Footer Style
    	modified:   src/components/Footer.css

    * Updating Footer Style again
    	modified:   src/components/Footer.css

    * Fixing white space
    	modified:   src/components/Footer.css
    	modified:   src/components/body.css
    	new file:   src/pages/LandingPage.tsx

    * put in new images & added image styling

    * Formatting into a page
    	modified:   src/App.tsx
    	modified:   src/pages/LandingPage.tsx

    * Image Refactor
    	modified:   src/App.css
    	deleted:    src/Images/unsplash_dsV4yvL5UJk.jpg
    	deleted:    src/components/Body.tsx
    	deleted:    src/components/body.css
    	modified:   src/pages/LandingPage.tsx

    * Landing Page_Body Refactor

    * Images
    	renamed:    src/components/Landing Page/Body.tsx -> src/components/Landing_Page/Body.tsx
    	renamed:    src/components/Landing Page/body.css -> src/components/Landing_Page/body.css
    	modified:   src/pages/LandingPage.tsx

    * Final Images
    	renamed:    src/components/Footer.css -> src/components/Footer/Footer.css
    	renamed:    src/components/Footer.tsx -> src/components/Footer/Footer.tsx
    	modified:   src/components/Landing_Page/Body.tsx
    	modified:   src/pages/LandingPage.tsx

    ---------

    Co-authored-by: meganleongg <98578642+meganleongg@users.noreply.github.com>

Anthonyp0329 committed:
Merge branch 'main' into Feature/vs2961/backend-auth

Anthonyp0329 committed:
SignUp Page styling

---- or ----, checkbox, and sign in with google button need to be styled/fixed still.

Anthonyp0329 committed:
changing around error display

Anthonyp0329 committed:
Update signup.css

Anthonyp0329 committed:
Update signup.css

Anthonyp0329 committed:
Merge branch 'main' into Feature/vs2961/backend-auth

Anthonyp0329 committed:
getting login, signup page routing wroking

vs2961 committed:
Update package.json

Anthonyp0329 committed:
Update index.js

Anthonyp0329 committed:
Update SignUpPage.tsx

navigates to home page after successful login

Anthonyp0329 committed:
Login page css

Anthonyp0329 committed:
SignUp page error handling

made error message pop up under corresponding input box

Anthonyp0329 committed:
LogIn Page error messages

Anthonyp0329 committed:
login page updates

Added "forgot your password" field, need to implement functionality. Added sign up for free part.

Anthonyp0329 committed:
changing class names

Anthonyp0329 committed:
Merge branch 'main' into Feature/vs2961/backend-auth

Anthonyp0329 committed:
changed signup button to look clickable

aammya8 committed:
Initial code for linking accounts

Anthonyp0329 committed:
Update auth.css

aammya8 committed:
Redirect to Landing Page after Google Login

aammya8 committed:
Improved Google Sign In Button CSS

aammya8 committed:
Changed Google Button CSS

aammya8 committed:
Fixed Google Button Styling

Anthonyp0329 committed:
arrow functions

Anthonyp0329 committed:
google button css

NirmalAgnihotri commented on SDCTA/src/components/auth/LogInPage.tsx:
```diff
@@ -0,0 +1,166 @@
+                loginEmail,
+                loginPassword
+            );
+            console.log(user);
```
Let's leave frontend logs out

NirmalAgnihotri commented on SDCTA/src/components/auth/LogInPage.tsx:
```diff
@@ -0,0 +1,166 @@
+    const [unknownError, setUnknownError] = useState("");
+
+    const login = async () => {
+        try {
```
So I actually think its ok to not tell them about a potential error in their password if their email is wrong. In other words you can just use the firebase error checking. 
But maybe a slightly more concise way to implement this would be to define an InputError type that is an object keyed with error type. So the keys would be strings like email, password, unknown . Then, instead of three states for the different errors you just have one state of type InputError that holds all the errors. Now, When you do this check you don't need if statements. You can just set the error with the key that matches the value of errorMessage[0] since your keys have the same name.

NirmalAgnihotri commented on SDCTA/src/components/auth/SignUpPage.tsx:
```diff
@@ -0,0 +1,291 @@
+
+      if (error instanceof Error) {
+        const errorMessage = await signUpErrorHandler(error);
+
```
Same thing with InputErrorType I mentioned above can be applied here, you can define that type in the auth-errors file you have.

NirmalAgnihotri commented on SDCTA/src/error_handling/auth-errors.tsx:
```diff
@@ -0,0 +1,57 @@
+const signUpErrorHandler = async (error: Error) => {
```
A slight variation on the InputError type I talked about above:
You could have these handlers return an InputError type instead of a tuple. You can do the same checks you're doing now in these functions but just set the correct key in the InputError object you're going to return when you find an error. Then return that InputError object. If there are no errors, the InputError object will just have keys with blank strings. Then instead of:

            setPasswordError("");
            setEmailError("");
            setUnknownError("");
            if (error instanceof Error) {
                const errorMessage = await logInErrorHandler(error);
                if (errorMessage[0] === "email") {
                    setEmailError(errorMessage[1]);
                }
                if (errorMessage[0] === "password") {
                    setPasswordError(errorMessage[1]);
                }
                if (errorMessage[0] === "unknown") {
                    setUnknownError(errorMessage[1] + " (Reload and try again)");
                }
            }

You can do something like:
if (error instanceOf Error) {
    setInputError(loginErrorHandler(error));
}

This is kind of rough off the top of my head, will prob have to account for some details when you implement it

NirmalAgnihotri commented on SDCTA/src/error_handling/auth-errors.tsx:
```diff
@@ -0,0 +1,57 @@
+    }
+}
+
+const logInErrorHandler = async (error: Error) => {
```
Do these handlers have to be async functions?

NirmalAgnihotri reviewed (changes requested):
Overall great work! Just left some suggestions on error handling.

vs2961 committed:
Merge branch 'main' into Feature/vs2961/backend-auth

Anthonyp0329 committed:
Fixed error handling, played around with css

Anthonyp0329 commented:
Changed error handling with your suggestions, should be better now!

Anthonyp0329 committed:
Redux state and reset password

Need to refine, but got reset password with email working
Also need to add sagas for function calls

Anthonyp0329 committed:
password reset improvements

Anthonyp0329 committed:
renaming conventions

Anthonyp0329 committed:
Merge branch 'main' into Feature/vs2961/backend-auth

NirmalAgnihotri committed:
Prettier and lint fixes

NirmalAgnihotri committed:
Linted and formatted backend files

NirmalAgnihotri committed:
Updated git and eslint ignore

NirmalAgnihotri committed:
lint fixes

NirmalAgnihotri committed:
fixed backend linting errors

Anthonyp0329 committed:
Linting Fixes

Login with google linting issues need to be fixed

Anthonyp0329 committed:
Update auth.css

NirmalAgnihotri committed:
lint cache in gitignore

NirmalAgnihotri commented on SDCTA/src/error_handling/auth-errors.tsx:
```diff
@@ -0,0 +1,79 @@
+interface inputError {
```
Small note, type names like classes in Java should start with a capital letter

NirmalAgnihotri replied:
Non-blocking, we can change this in a future PR

NirmalAgnihotri commented on SDCTA/src/components/auth/SignUpPage.tsx:
```diff
@@ -0,0 +1,296 @@
+        navigate("/");
+      })
+      .catch((error) => {
+        // const errorCode = error.code;
```
Do we still need this commented out code? If so, that's ok but lets try to remove it in a future auth PR

NirmalAgnihotri reviewed (approved):
Great job implementing the changes! Just fix the linting issues and push. I left some minor comments, if while fixing the linting issues you can just capitalize the type names that would be great. Otherwise just do it in the next auth PR.

Anthonyp0329 committed:
fixing linter issues

fixing linter issues

Anthonyp0329 committed:
Update .eslintrc.json

aammya8 committed:
fixing linting errors

aammya8 committed:
Merge branch 'Feature/vs2961/backend-auth' of https://github.com/TritonSE/SDCTA-Data-Visualization-Website into Feature/vs2961/backend-auth

aammya8 committed:
fixing linter errors

vs2961 committed:
Merge branch 'main' into Feature/vs2961/backend-auth

vs2961 committed:
Fix Merge Conflictsg

vs2961 committed:
Fix linting

aammya8 committed:
Fixed Linting Error

aammya8 committed:
Fixed Linting Error

aammya8 committed:
Fixed linting error

aammya8 committed:
update redux state

aammya8 committed:
Fixed redirect, linting errors

Do not redirect from google sign in unless authenticated

aammya8 committed:
Fixed Linting errors

aammya8 committed:
Removed unnecessary comments

NirmalAgnihotri committed:
Prettier fixes
